Upgrade to JUnit Jupiter

mureinik opened the pull request:
This PR follows commons-lang's example and migrates the project to the modern JUnit Jupiter 5.4.1 testing framework.
Since JUnit Jupiter is not backwards compatible to JUnit 4.x, or even JUnit Vintage 5.x, the patch is a tad large, although most of the changes are just cosmetic.

1. Maven changes:
    1. `junit:junit` was replaced with `org.junit.jupiter:junit-jupiter`.
    2. `org.hamcrest:hamcrest-all` was added since JUnit Jupiter no longer bundles Hamcrest with them.

2. Annotations:
   1. `org.junit.jupiter.api.Test` was used as a drop in replacement for `org.juit.Test` without arguments. See 3.ii for handling of the`@Test` annotation with an `expected` argument.
   2. `org.junit.jupiter.api.BeforeEach` was used as an drop in replacement for `org.junit.Before`.
   3. `org.junit.jupiter.api.AfterEach` was used as a drop in replacement for `org.junit.After`.
   4. `org.junit.jupiter.api.BeforeAll` was used as a drop in replacement for `org.junit.BeforeClass`.
   5. `org.junit.jupiter.api.AfterAll` was used as a drop in replacement for `org.junit.AfterClass`.
   6. `org.junit.jupiter.api.Disabled` was used as a drop in replacement for `org.junit.Ignore`.

3. Assertions:
   1. `org.junit.jupiter.api.Assertions`' methods were used as a drop in replacements for `org.junit.Assert`'s methods with the same name in the simple case of an assertion without a message.  In the case of an assertion with a message, `org.junit.jupiter.api.Assertions`' methods were used, but the argument order was changed - `Assert`'s methods take the message as the first argument, while `Assertions`' methods take the message as the last argument.
   2. `org.junit.jupiter.api.Assertions#assertThrows` was used to assert a specific exception was thrown instead of an `org.junit.Test` annotation with an `expected` argument. This technique has a side bonus of making the tests slightly stricter, as now they assert the exception was thrown from a specific line and prevent false posivites where the test's "set-up" code accidentally threw that exception. The `throws` clauses of these methods were cleaned up from exceptions that can no longer be thrown in order to avoid compilation warnings.
   3. `org.hamcrest.MatcherAssert#assertThat` was used as a drop in reaplcement for `org.junit.Assert#assertThat`, as `org.junit.jupiter.api.Assertions` has no equivilent method.

garydgregory commented:
Hi @mureinik,
Thank you for your PR. May you please rebase so we can get these conflicts resolved? 
Thank you!

mureinik commented:
@garydgregory sure, done.
WIth this update to the PR, I included the following:
- Rebased and resolved merge-conflicts
- Converted `TestParallelCreationWithNoIdle` to JUnit Jupiter, as this test was added after the original PR was authored
- Upgraded to JUnit Jupiter 5.4.1, which is the latest available version (only 5.3.1 was available when the original PR was authored)
- Used the simplified aggregate dependency `org.junit.jupiter:junit-jupiter` offered by 5.4 to make the patch a tad neated.

mureinik commented:
@garydgregory regarding the build failures - they are not related to this patch. They are related to newer JDKs (11 and ea), and the same failures occur on master too.
